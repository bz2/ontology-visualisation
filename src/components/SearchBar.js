import { useState } from 'react'
import { connect } from 'redux-zero/react'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'
import { AutoComplete } from 'primereact/autocomplete'
import { ProgressSpinner } from 'primereact/progressspinner'
import { Button } from 'primereact/button'
import actions from '../store/actions'
import getSuggestions from '../utils/graphSearch/getSuggestions'
import searchGraph from '../utils/graphSearch/searchGraph'
import { OPERATION_TYPE_UPDATE } from '../constants/store'

const SearchBar = ({
  updateStoreValue,
  entrySearchValue,
  isSearchLoading
}) => {
  const { t } = useTranslation()

  const [suggestions, setSuggestions] = useState([])

  return (
    <>
      {
        isSearchLoading
          ? (
            <div className="graph-search-loader">
              <ProgressSpinner
                className="spinner"
                strokeWidth="4"
              />
            </div>
          ) : (
            <div className="graph-search-bar">
              <form onSubmit={(e) => {
                e.preventDefault()

                updateStoreValue(['searchPageSelected'], OPERATION_TYPE_UPDATE, 0)

                searchGraph({
                  updateStoreValue,
                })
              }}
              >
                <AutoComplete
                  name="search"
                  value={entrySearchValue}
                  id="main-search"
                  suggestions={suggestions}
                  completeMethod={(e) => {
                    const { query } = e

                    getSuggestions({
                      query,
                      setSuggestions,
                      updateStoreValue,
                      t
                    })
                  }}
                  field="label"
                  autoHighlight
                  scrollHeight="200px"
                  delay={500}
                  onSelect={(e) => {
                    const { value } = e.value

                    updateStoreValue(['searchPageSelected'], OPERATION_TYPE_UPDATE, 0)
                    updateStoreValue(['entrySearchValue'], OPERATION_TYPE_UPDATE, value)

                    searchGraph({
                      updateStoreValue,
                      t
                    })
                  }}
                  onChange={(e) => updateStoreValue(['entrySearchValue'], OPERATION_TYPE_UPDATE, e.value)}
                />
                <Button
                  icon="pi pi-search"
                  id="search-icon"
                  className="p-button-rounded"
                  onClick={() => {
                    updateStoreValue(['searchPageSelected'], OPERATION_TYPE_UPDATE, 0)

                    searchGraph({
                      updateStoreValue,
                      t
                    })
                  }}
                />
                <input
                  name="submit"
                  className="hidden"
                  type="submit"
                />
              </form>
            </div>
          )
        }
    </>
  )
}

SearchBar.propTypes = {
  updateStoreValue: PropTypes.func.isRequired,
  entrySearchValue: PropTypes.string.isRequired,
  isSearchLoading: PropTypes.bool.isRequired,
}

const mapToProps = ({
  entrySearchValue,
  isSearchLoading
}) => ({
  entrySearchValue,
  isSearchLoading
})

export default connect(
  mapToProps,
  actions
)(SearchBar)
