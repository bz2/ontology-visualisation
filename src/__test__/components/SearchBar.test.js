import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import SearchBar from '../../components/SearchBar'

const setup = ({
  isSearchLoading
}) => {
  const props = {
    updateStoreValue: jest.fn(),
    entrySearchValue: 'road',
    isSearchLoading
  }

  const component = shallow(<SearchBar {...props} />)

  return {
    component,
    props
  }
}

describe('SearchBar', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot when loading', () => {
    const {
      component
    } = setup({
      isSearchLoading: true
    })

    expect(toJson(component)).toMatchSnapshot()
  })

  it('should match snapshot when not loading', () => {
    const {
      component
    } = setup({
      isSearchLoading: false
    })

    expect(toJson(component)).toMatchSnapshot()
  })
})
