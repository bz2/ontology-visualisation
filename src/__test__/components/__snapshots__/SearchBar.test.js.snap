// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot when loading 1`] = `
<Connect
  actions={
    Object {
      "updateStoreValue": [Function],
    }
  }
  entrySearchValue="road"
  isSearchLoading={true}
  mapToProps={[Function]}
  updateStoreValue={[MockFunction]}
>
  <Component />
</Connect>
`;

exports[`SearchBar should match snapshot when not loading 1`] = `
<Connect
  actions={
    Object {
      "updateStoreValue": [Function],
    }
  }
  entrySearchValue="road"
  isSearchLoading={false}
  mapToProps={[Function]}
  updateStoreValue={[MockFunction]}
>
  <Component />
</Connect>
`;
